### ISSUE TITLE
Question use of glog

### ISSUE THREAD

**Issue by kris-nova** [kubernetes/kops#1013]
When using kops as a library, glog is a PITA to implement and work with.

The constructor calls `flag` directly in glog, and if you want to implement other command line flags in go (EG Cobra) there is some hacky things that need to be done..

I know `glog` is the official Google logging tool, but wondering if we can entertain the idea of another logger in kops.

Thoughts?

CC @chrislovecnm @justinsb

**Comment by aledbf** [comment]
@kris-nova I think you is not possible because glog is used everywhere in k8s so even if you manage to find a replacement in kops you will face the same issue

**Comment by justinsb** [comment]
I agree with the sentiment (though not necessarily the reasoning around flags); I would like to get the following:

1) prettier CLI output (e.g. colorized?)
2) capture output when doing server operations (but go essentially makes this impossible, because of a lack of thread-locals or thread-ids).

However, I think @aledbf is right - it doesn't do us as much good if we just do it in our code (particularly as we use more and more of the k8s API machinery), and we should coordinate with kubernetes/kubernetes.  @thockin has looked into it, but not sure if there's any good options.  I don't mind taking the lead and trying something new, but I don't want to diverge from kubernetes/kubernetes if they're going in a different direction.

**Comment by thockin** [comment]
glog is fine, for what it is, but it isn't particularly good or pretty.
What I REALLY dislike is that code just calls `glog.Something()`, causing
it all to depend on glog.  Our discipline around logging has been very
poor.  We should have defined a simple interface and passed instances of
that around to the places that need logging, and nobody but main()-ish
functions should depend on glog.

I also dislike the myriad extra things glog does.  I offer
https://github.com/thockin/logr as a simpler API to logging, and
https://github.com/thockin/glogr as an implementation.

That said, I also think structured logging is super cool, and I think we
should try it out.  I like the look of https://github.com/uber-go/zap but I
have not used it extensively.

I don't really care about color, frankly.  I am fine with it, but I
wouldn't go out of my way to get it.

So if I had my druthers:
a) find a blend of logr and structured logging
b) implement in terms of zap
c) start the business of converting modules-by-module to use this new,
injected logger
d) kill off glog (or abandon structured logs and keep glog via glogr)

We need a testbed for this.  I used logr in a few smaller repos (e.g.
spartakus) and I thought it was great, but I usually think the things I do
are great, so I want someone ELSE to tell me it is great.  Or that is is
not, and we can retool.

This will be a MASSIVE but mostly mechanical undertaking.

**Comment by brandoncole** [comment]
I would 👎  on colorized unless it's optional.  A lot of time the color encodings - ie. `\033[0;31m` if that's what we're talking about - show up as plain text and not colors in log collection systems if this task is automated.

**Comment by chrislovecnm** [comment]
Great point
